### ISSUE TITLE
Eloquent builders for scopes support

### ISSUE THREAD

**Issue by stayallive** [nunomaduro/larastan#918]
- Larastan Version: 0.7.12
- `--level` used: 2

### Description

As far as I understand it to have Larastan correctly understand scopes it is required to use a custom builder, or at least this used to be true I believe. But I'm having trouble getting it to not throw errors on the builder class.

The custom builder class gives this error:

```
phpstan: Type mixed in generic type Illuminate\Database\Eloquent\Builder<mixed> in PHPDoc tag @extends is not subtype of template type TModelClass of Illuminate\Database\Eloquent\Model of class Illuminate\Database\Eloquent\Builder.
```

I used this to get to the PHPDoc I have on that class: https://github.com/nunomaduro/larastan/blob/master/UPGRADE.md#custom-eloquent-builders

The builder class (with the scope methods removed for brevity):

```php
/**
 * @template TModelClass
 * @extends \Illuminate\Database\Eloquent\Builder<TModelClass>
 */
class UserBuilder extends Builder
{}
```

My Model:

```php
    /**
     * @param \Illuminate\Database\Query\Builder $query
     *
     * @return \Simulise\Models\QueryBuilders\UserBuilder<\Simulise\Models\User>
     */
    public function newEloquentBuilder($query): UserBuilder
    {
        return new UserBuilder($query);
    }
    public static function query(): UserBuilder
    {
        return parent::query();
    }
```

(The `query()` override is not needed for Larastan but it is for PhpStorm to play nice too :))

So long story short: The custom query builder works, and Larastan is able to get what it's doing and resolve the scopes but is generating an error on the custom builder class itself.

Hope it did not miss something stupid but apart from the upgrade guide there is no documentation about how to get scopes to work 😄

**Comment by canvural** [comment]
Hi,

> As far as I understand it to have Larastan correctly understand scopes it is required to use a custom builder

No, this was never the case. Scopes just work without any extra config.

Maybe you are getting the error because you did not specify the generic parameters for the `query` method return type. Try this:

```php
/**
 * @return \Simulise\Models\QueryBuilders\UserBuilder<\Simulise\Models\User>
 * /
public static function query(): UserBuilder
{
    return parent::query();
}
```

**Comment by stayallive** [comment]
Interesting, and you are right... in that case I'm going to take the easy route and simply remove the custom builder since it's also not required for PhpStorm with the right plugin!

I tried your suggestion to test it out but the error remains the same. But I would be happy to say that's user error and since I'm removing the builder anyway not put much time in getting to the bottom of it :)

**Comment by canvural** [comment]
I just tested something similar [here.](https://phpstan.org/r/e289aa14-da5c-441f-986e-5d4d9ceffb5e) I think this is the problem. So custom builder should be defined like this maybe:

/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 * @extends \Illuminate\Database\Eloquent\Builder<TModelClass>
 */
class UserBuilder extends Builder
{}

**Comment by stayallive** [comment]
I tried that too and it does fix the error now, nice 👍

**Comment by canvural** [comment]
Alright then, we need to fix it in the upgrade guide.

Thank you.

**Comment by stayallive** [comment]
Thanks for the quick response by the way, much appreciated 👍
